File: aoa_hw2/aoa_hw2/decode.py
from cStringIO import StringIO
import sys

def main():
    if len(sys.argv) != 4:
        print "Usage : python decode.py encoded.txt hcodes.txt decoded.txt"
        return

    f = open(sys.argv[2])
    lines = f.readlines()
    f.close()
    code_dict = {}
    for line in lines:
        [value, key] = line.strip().split()
        if value == 'space':
            value = ' '
        code_dict[key] = value



    f = open(sys.argv[1])
    lines = f.readline()
    f.close()
    l = len(lines)

    org_text = StringIO()
    i,j = 0,1
    while j <= l:
        word = lines[i:j]
        if code_dict.has_key(word):
            org_text.write(code_dict[word])
            i = j
        j = j + 1
        
    if i != l:
        print 'Error in decoding!!'

    f = open(sys.argv[3], 'w')
    f.write(org_text.getvalue())

if __name__ == "__main__":
    main()

File: aoa_hw2/decode.cpp
#include<iostream>
#include<fstream>
#include<string>
#include<map>
#include<sstream>

using namespace std;

int main(int argc, char *argv[]) {

	if(argc != 4){
		cout << "Usage : decode encoded.txt hcodes.txt decoded.txt" << endl;
		return -1;
	}

	fstream f;
	string line, word;
	map<string, char> code_dict;
	map<string, char>::iterator it;
	int i, j, l;
	ostringstream org_text;

	f.open(argv[2], ios::in);
	while(!f.eof()) {
		getline(f, line);
		if(line.length() == 0)
			break;
		if(line[1] == ' ')
			code_dict[line.substr(2)] = line[0];
//			code_dict.insert(make_pair(line.substr(2), line[0]));
		else
//			code_dict.insert(make_pair(line.substr(6), ' '));
			code_dict[line.substr(6)] = ' ';
	}
	f.close();

//	for(it = code_dict.begin(); it != code_dict.end(); ++it)
//		cout << it->first << ' ' << it->second << endl;


	f.open(argv[1], ios::in);
	getline(f, line);
	f.close();

	l = line.length();
	i = 0; j = 1;
	while(j <= l) {
		word = line.substr(i, j - i);
		if((it = code_dict.find(word)) != code_dict.end()) {
			org_text << it->second;
			i = j;
		}
		j++;
	}
	if(i != l)
		cout << "Error in decoding!!!";

	f.open(argv[3], ios::out);
	f << org_text.str();
	f.close();

}
